src/testrender/shading.cpp

```diff
@@ -860,6 +1063,92 @@ process_closure(ShadingResult& result, const ClosureColor* closure,
+                    ok = result.bsdf.add_bsdf<Microfacet<BeckmannDist, 2>, MicrofacetParams>(cw, params);
+                break;
+            }
+            case MX_CONDUCTOR_ID: {
+                const MxConductorParams* srcparams = comp->as<MxConductorParams>();
+                MicrofacetParams params = {};
+                params.N = srcparams->N;
+                params.U = srcparams->U;
+                params.dist = srcparams->distribution;
+                params.xalpha = srcparams->roughness_x;
+                params.yalpha = srcparams->roughness_y;
+                params.eta = 100.0f; // TODO: support conductor fresnel equation properly
+                if (params.dist == u_ggx)
+                    ok = result.bsdf.add_bsdf<Microfacet<GGXDist, 0>, MicrofacetParams>(cw, params);
+                else if (params.dist == u_beckmann)
+                    ok = result.bsdf.add_bsdf<Microfacet<BeckmannDist, 0>, MicrofacetParams>(cw, params);
+            };
+            case MX_GENERALIZED_SCHLICK_ID: {
+                const MxGeneralizedSchlickParams* srcparams = comp->as<MxGeneralizedSchlickParams>();
+                MicrofacetParams params = {};
+                params.N = srcparams->N;
+                params.U = srcparams->U;
+                params.dist = srcparams->distribution;
+                params.xalpha = srcparams->roughness_x;
+                params.yalpha = srcparams->roughness_y;
+                params.eta = 100.0f; // TODO: support generalized fresnel equation properly
+                // TODO: figure out when the material is meant to transmit light
+                if (params.dist == u_ggx)
+                    ok = result.bsdf.add_bsdf<Microfacet<GGXDist, 0>, MicrofacetParams>(cw, params);
+                else if (params.dist == u_beckmann)
+                    ok = result.bsdf.add_bsdf<Microfacet<BeckmannDist, 0>, MicrofacetParams>(cw, params);
+            };
+            case MX_TRANSLUCENT_ID: {
+                const MxTranslucentParams* srcparams = comp->as<MxTranslucentParams>();
+                DiffuseParams params = {};
+                params.N = srcparams->N;
+                ok = result.bsdf.add_bsdf<Diffuse<1>, DiffuseParams>(cw * srcparams->albedo, params);
+                break;
+            }
+            case MX_TRANSPARENT_ID: {
+                ok = result.bsdf.add_bsdf<Transparent, int>(cw, 0);
+                break;
+            }
+            case MX_SUBSURFACE_ID: {
+                // TODO: implement BSSRDF support?
+                const MxSubsurfaceParams* srcparams = comp->as<MxSubsurfaceParams>();
+                DiffuseParams params = {};
+                params.N = srcparams->N;
+                ok = result.bsdf.add_bsdf<Diffuse<0>, DiffuseParams>(cw * srcparams->albedo, params);
+                break;
+            }
+            case MX_SHEEN_ID:
+            case MX_LAYER_ID: {
+                OSL_ASSERT(false && "MaterialX closure not yet implemented");
```

Comment:
I have a suggestion on how to implement this. Add a 

virtual Color3 filter_base(Vec3 wo);

You call it here, scale the weight of base and just add_bsdf() both. Recurse where needed. Works for us by making assumptions on wi. Eliminates the need of recursion at eval/sample time. De-virtualize and you are ready for GPU.

Reply:
Thats what I was thinking too. Maybe it can even just use the existing "get_albedo()" method?

Reply:
Good idea, although the coating may want to add absorption according to some thickness. So maybe a different method that by default does 1 - get_albedo().